### DISCUSSION TITLE
JWKS Keys not found

### DISCUSSION THREAD

**Discussion by sevensolutions** [zitadel/zitadel#7961] [Q&A]
Hi,
does anyone else have some problems using ZITADEL with .NET?
I configured it using the standard .AddOpenIdConnect() method and everything is working fine after starting the app.
But after a few hours i can't login anymore, because it cant verify my token anymore, because it looks like ZITADEL has rotated the JWKS.

The default cache duration in .NET is 12h and i thought: Well, there is a configuration in ZITADEL which says "PrivateKeyLifetime" is 6h.
So i changed this to 24h but this also doesn't work. It looks like the keys are still rotated within 12h.

It sounds like a bug to me or is there anything i'am missing in configuration?

As soon as i refresh the JWKS cache or restart the app, it's immetiately working again.
It looks like ZITADEL is rotating the key and immediately using the new keys to sign the tokens. Shouldn't there be a delay so that the apps can update their caches?

**Comment by sbosor** [comment]
Hi! 

I am also experiencing the same and I have also tried modified the cookie duration in .Net and the PrivateKeyLifetime in Zitadel, but the issue still reproduces. As you have noticed, I have [created an issue](https://github.com/smartive/zitadel-net/issues/927) describing the behavior and also wrote on the Zitadel's Discord channels, but I did not receive any solution or recommendations on what to do so far.

Another problem I had with it is that I  could not simulate the behavior and debug it locally. How did you refresh the JWKS cache?

**Reply by sevensolutions** [reply]
I just restarted my app to reset the cache.

I'am using the following workaround until we have a fix:
```csharp
.AddOpenIdConnect( options =>
{
	options.Authority = builder.Configuration.GetValue<string>( "Authentication:Authority" );
	// Rest has been ommitted
	
	var jwksUri = builder.Configuration.GetValue<string>( "Authentication:JwksUri" );
	if ( jwksUri is not null )
	{
		options.TokenValidationParameters.IssuerSigningKeyResolver = ( _, _, kid, _ ) =>
		{
			using var httpClient = new HttpClient();

			var jsonAsync = httpClient.GetStringAsync( jwksUri );

			return JsonWebKeySet.Create( jsonAsync.GetAwaiter().GetResult() )
				.GetSigningKeys()
				.Where( x => x.KeyId == kid )
				.ToArray();
		};
	}
} );
```
It's not ideal because it's refreshing the keys to often which may open the app to DDOS attacks.

Maybe i will open a real issue in the ZITADEL repo but i wanted to write here first, to ensure it's not my fault just because of a bad misconfiguration.

**Reply by sevensolutions** [reply]
Just an update, because i found some more infos on this:
It looks like the OIDC Middleware in .NET Core already supports automatic key rollover.
https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.openidconnect.openidconnectoptions.refreshonissuerkeynotfound?view=aspnetcore-8.0

This setting is enabled by default.

But it also looks like there is an open bug:
https://github.com/dotnet/aspnetcore/issues/28948

I can remember that it sometimes started to work again, if i tried a few consecutive logins.

**Reply by sbosor** [reply]
Thanks for the explanation and for the update, @sevensolutions! Indeed, sometimes it eventually works after trying a few consecutive logins, but other times it doesn't.

**Comment by hifabienne** [comment]
@livio-a @muhlemmer can you help here?

**Comment by muhlemmer** [comment]
@sevensolutions your feedback is similar we've received already on keys management. This thread explains the current problems we see with the keys and the proposed solution: https://github.com/zitadel/zitadel/discussions/7464.

This is the enhancement issue for implementation: https://github.com/zitadel/zitadel/issues/7809

**Reply by sevensolutions** [reply]
@muhlemmer thanks for clarification. It's no problem for me to wait for a solution.
In the meantime i'am using my workaround.
